File: problems/152/solution.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ans = nums[0], maxSoFar = nums[0], minSoFar = nums[0];
        
        for (int i = 1; i < nums.size(); i++) {
            int cur = nums[i];
            
            int tmp = max({cur, maxSoFar * cur, minSoFar * cur});
            minSoFar = min({cur, maxSoFar * cur, minSoFar * cur});
            
            maxSoFar = tmp;
            ans = max(ans, maxSoFar);
        }
        
        return ans;
    }
};

File: problems/152/solution.go
package main
func maxProduct(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    
    ans := nums[0]
    
    for i := 0; i < len(nums); i++ {
        cur := 1
        
        for j := i; j < len(nums); j++ {
            cur *= nums[j]
            
            if cur > ans {
                ans = cur
            }
        }
    }
    
    return ans
}